src/daemon/distribution.c

```diff
@@ -0,0 +1,258 @@
+ */
+
+static size_t left_child_index(size_t node_index,
+                               __attribute__((unused)) size_t left,
+                               __attribute__((unused)) size_t right) {
+  return node_index + 1;
+}
+
+static size_t right_child_index(size_t node_index, size_t left, size_t right) {
+  size_t mid = (left + right) / 2;
+  return node_index + 2 * (mid - left + 1);
+}
+
+static size_t tree_size(size_t num_buckets) {
+  return 2 * num_buckets - 1;
+}
+
+static bucket_t merge_buckets(bucket_t left_child, bucket_t right_child) {
+  return (bucket_t) {
+      .bucket_counter = left_child.bucket_counter + right_child.bucket_counter,
+      .minimum = left_child.minimum,
+      .maximum = right_child.maximum,
+  };
+}
+
+static void build_tree(distribution_t *d, bucket_t *buckets, size_t node_index, size_t left, size_t right) {
+  if (left > right)
+    return;
+  if (left == right) {
+    d->tree[node_index] = buckets[left];
+    return;
+  }
+  size_t mid = (left + right) / 2;
+  size_t left_child = left_child_index(node_index, left, right);
+  size_t right_child = right_child_index(node_index, left, right);
+  build_tree(d, buckets, left_child, left, mid);
+  build_tree(d, buckets, right_child, mid + 1, right);
+  d->tree[node_index] = merge_buckets(d->tree[left_child], d->tree[right_child]);
+}
+
+static distribution_t* build_distribution_from_bucket_array(size_t num_buckets, bucket_t *bucket_array) {
+  distribution_t *new_distribution = calloc(1, sizeof(distribution_t));
+  if (new_distribution == NULL) {
+    return NULL;
+  }
+  new_distribution->tree = calloc(tree_size(num_buckets), sizeof(bucket_t));
+  if (new_distribution->tree == NULL) {
+    free(new_distribution);
+    return NULL;
+  }
+
+  new_distribution->num_buckets = num_buckets;
+  build_tree(new_distribution, bucket_array, 0, 0, num_buckets - 1);
+  return new_distribution;
+}
+
+distribution_t* distribution_new_linear(size_t num_buckets, double size) {
+  if (num_buckets == 0 || size <= 0) {
+    errno = EINVAL;
+    return NULL;
+  }
+
+  bucket_t bucket_array[num_buckets];
+  for (size_t i = 0; i < num_buckets; i++) {
+    bucket_array[i].bucket_counter = 0;
+    bucket_array[i].minimum = i * size;
+    if (i == num_buckets - 1)
+      bucket_array[i].maximum = INFINITY;
+    else
+      bucket_array[i].maximum = (i + 1) * size;
+  }
+  return build_distribution_from_bucket_array(num_buckets, bucket_array);
+}
+
+distribution_t* distribution_new_exponential(size_t num_buckets, double initial_size, double factor) {
+  if (num_buckets == 0 || initial_size <= 0 || factor <= 1) {
+    errno = EINVAL;
+    return NULL;
+  }
+
+  bucket_t bucket_array[num_buckets];
+  bucket_array[0] = (bucket_t) {
+    .bucket_counter = 0,
+    .minimum = 0,
+    .maximum = initial_size,
+  };
+  for (size_t i = 1; i < num_buckets; i++) {
+    bucket_array[i].bucket_counter = 0;
+    bucket_array[i].minimum = bucket_array[i - 1].maximum;
```

Comment:
I implemented such approach and realised that now my constructor functions are very similar. The main difference is in calculating maximum boundaries of buckets. Maybe, it would be better to select maximum boundaries first and then call the function that will fill other fields? Or such code duplication is okay?